qutebrowser/browser/pdfjs.py

```diff
@@ -0,0 +1,151 @@
+from qutebrowser.utils import utils
+
+
+class PDFJSNotFound(Exception):
+
+    """Raised when no pdf.js installation is found."""
+
+    pass
+
+
+def generate_pdfjs_page(url):
+    """Return the html content of a page that displays url with pdfjs.
+
+    Returns a string.
+
+    Args:
+        url: The url of the pdf as QUrl.
+    """
+    viewer = get_pdfjs_res('web/viewer.html').decode('utf-8')
+    script = _generate_pdfjs_script(url)
+    html_page = viewer.replace(
+        '</body>', '</body><script>{}</script>'.format(script)
+    )
+    return html_page
+
+
+def _generate_pdfjs_script(url):
+    """Generate the script that shows the pdf with pdf.js.
+
+    Args:
+        url: The url of the pdf page as QUrl.
+    """
+    return (
+        'PDFJS.getDocument("{url}").then(function(pdf) {{\n'
+        '    PDFView.load(pdf);\n'
+        '}});'
+    ).format(url=webelem.javascript_escape(url.toString()))
+
+
+def fix_urls(asset):
+    """Take a html page and replace each relative URL wth an absolute.
+
+    This is specialized for pdf.js files and not a general purpose function.
+
+    Args:
+        asset: js file or html page as string.
+    """
+    new_urls = {
+        'viewer.css': 'qute://pdfjs/web/viewer.css',
+        'compatibility.js': 'qute://pdfjs/web/compatibility.js',
+        'locale/locale.properties':
+            'qute://pdfjs/web/locale/locale.properties',
+        'l10n.js': 'qute://pdfjs/web/l10n.js',
+        '../build/pdf.js': 'qute://pdfjs/build/pdf.js',
+        'debugger.js': 'qute://pdfjs/web/debugger.js',
+        'viewer.js': 'qute://pdfjs/web/viewer.js',
+        'compressed.tracemonkey-pldi-09.pdf': '',
+        './images/': 'qute://pdfjs/web/images/',
+        '../build/pdf.worker.js': 'qute://pdfjs/build/pdf.worker.js',
+        '../web/cmaps/': 'qute://pdfjs/web/cmaps/',
+    }
+    for original, new in new_urls.items():
+        asset = asset.replace(original, new)
+    return asset
+
+
+SYSTEM_PDFJS_PATHS = [
+    '/usr/share/pdf.js/',  # Debian pdf.js-common
+    '/usr/share/javascript/pdf/',  # Debian libjs-pdf
+    os.path.expanduser('~/.local/share/qutebrowser/pdfjs/'),  # fallback
+]
+
+
+def get_pdfjs_res(path):
+    """Get a pdf.js resource in binary format.
+
+    Args:
+        path: The path inside the pdfjs directory.
+    """
+    path = path.lstrip('/')
+    content = None
+
+    # First try a system wide installation
+    # System installations might strip off the 'build/' or 'web/' prefixes.
+    # qute expects them, so we need to adjust for it.
+    names_to_try = [path, path[path.index('/')+1:]]
+    for system_path in SYSTEM_PDFJS_PATHS:
+        content = _read_from_system(system_path, names_to_try)
+        if content is not None:
+            break
+
+    # Fallback to bundled pdf.js
+    if content is None:
+        res_path = '3rdparty/pdfjs/{}'.format(path)
+        try:
+            content = utils.read_file(res_path, binary=True)
+        except FileNotFoundError:
```

Comment:
Maybe `OSError` should be handled here in general? Not sure about it though... But I think if an user has a pdfjs available and, say, unreadable, they should get the error page as well.